mesonbuild/dependencies/mpi.py

```diff
@@ -37,65 +38,79 @@ def mpi_factory(env: 'Environment',
         return []
     compiler_is_intel = compiler.get_id() in {'intel', 'intel-cl'}
 
-    # Only OpenMPI has pkg-config, and it doesn't work with the intel compilers
-    if DependencyMethods.PKGCONFIG in methods and not compiler_is_intel:
-        pkg_name = None
+    if DependencyMethods.CONFIG_TOOL in methods:
+        nwargs = kwargs.copy()
+
+        # We try the environment variables for the tools first, but then
+        # fall back to the hardcoded names
+
         if language == 'c':
-            pkg_name = 'ompi-c'
+            env_vars = ['MPICC']
         elif language == 'cpp':
-            pkg_name = 'ompi-cxx'
+            env_vars = ['MPICXX']
         elif language == 'fortran':
-            pkg_name = 'ompi-fort'
-        candidates.append(functools.partial(
-            PkgConfigDependency, pkg_name, env, kwargs, language=language))
+            env_vars = ['MPIFC', 'MPIF90', 'MPIF77']
 
-    if DependencyMethods.CONFIG_TOOL in methods:
-        nwargs = kwargs.copy()
+        tool_names = [os.environ.get(env_name) for env_name in env_vars]
```

Comment:
Sigh. I know what this module is doing, but if you'd like to be extra awesome, we really want to do all of these environment variable lookups in the `Environment` class, which would add the ability to look these components up from a machine file, as well as keep our guarantee that once configured Meson won't re-read and from the environment and thus change the configuration behind your back on a reconfigure.

If you'd rather not you can leave it and I'll follow up with that cleanup

Reply:
> If you'd rather not you can leave it and I'll follow up with that cleanup

I understand your point but I don't know how to do this in Meson, so yes it would be more efficient if you follow up.